### PR TITLE
feat(notched-outline): typescript conversion

### PR DESCRIPTION
**Pull request by gugu**
Upgrade to v1 and fix typing issues

### CODE REVIEW

File: packages/notched-outline/index.tsx [hunk 1]
```diff
@@ -122,6 +121,7 @@ export default class NotchedOutline extends React.Component<
         <div
           className={this.classes}
           key='notched-outline'
+          style={ {width: this.state.actualWidth ? `${this.state.actualWidth}px`: undefined} }
```
Comment (moog16) [inline packages/notched-outline/index.tsx]: The render method needs to change quite a bit. I think with the new notchedoutline, it requires 3 divs

**Comment by moog16** [comment]
@gugu if you want to focus here - I can focus on the select updates

**Comment by gugu** [comment]
@moog16 yesterday I started with ripple/notched outline. Then I can continue with components you choose

**Comment by gugu** [comment]
It looks ugly in screenshot, because it needs floating label & textfield update to work correctly

File: packages/notched-outline/index.tsx [hunk 2]
```diff
@@ -91,49 +89,44 @@ export default class NotchedOutline extends React.Component<
...
-        />
-      </React.Fragment>
+      <div
+        className={this.classes}
+        key='notched-outline'
+        style={ {width: this.state.actualWidth ? `${this.state.actualWidth}px`: undefined} }
+        ref={this.outlineElement_}
+      >
+        <div className='mdc-notched-outline__leading'></div>
+        {this.props.children ?
```
Comment (moog16) [inline packages/notched-outline/index.tsx]: This should actually be:
```
<div className='mdc-notched-outline__notch'>
  {this.props.children}
</div>
```

We should always show the __notch element.

File: packages/notched-outline/index.tsx [hunk 3]
```diff
@@ -91,49 +89,44 @@ export default class NotchedOutline extends React.Component<
...
-        <div
-          ref={this.idleElement_}
-          className='mdc-notched-outline__idle'
-          key='notched-outline-idle'
-        />
-      </React.Fragment>
+      <div
+        className={this.classes}
+        key='notched-outline'
+        style={ {width: this.state.actualWidth ? `${this.state.actualWidth}px`: undefined} }
```
Comment (moog16) [inline packages/notched-outline/index.tsx]: - Pull this out into a variable
- Allow for users to apply their own styles (this.props.styles), but have our width override theirs

File: packages/notched-outline/index.tsx [hunk 4]
```diff
@@ -91,49 +89,44 @@ export default class NotchedOutline extends React.Component<
...
-            <path ref={this.pathElement_} className='mdc-notched-outline__path' />
-          </svg>
-        </div>
-        <div
-          ref={this.idleElement_}
-          className='mdc-notched-outline__idle'
-          key='notched-outline-idle'
-        />
-      </React.Fragment>
+      <div
```
Comment (moog16) [inline packages/notched-outline/index.tsx]: Not sure what otherProps users would add, but please add a way to pass properties here.

File: packages/notched-outline/index.tsx [hunk 5]
```diff
@@ -91,49 +89,44 @@ export default class NotchedOutline extends React.Component<
...
+        className={this.classes}
+        key='notched-outline'
```
Comment (moog16) [inline packages/notched-outline/index.tsx]: is this key necessary?

File: packages/notched-outline/index.tsx [hunk 6]
```diff
@@ -21,46 +21,45 @@
...
+  actualWidth: number | null,
 };
 
 export default class NotchedOutline extends React.Component<
   NotchedOutlineProps,
   NotchedOutlineState
   > {
-  foundation_: MDCNotchedOutlineFoundation;
+  foundation_!: MDCNotchedOutlineFoundation;
   outlineElement_: React.RefObject<HTMLDivElement> = React.createRef();
```
Comment (moog16) [inline packages/notched-outline/index.tsx]: This doesn't need to be here anymore

File: test/unit/notched-outline/index.test.tsx [hunk 7]
```diff
@@ -3,29 +3,31 @@ import {assert} from 'chai';
 import * as td from 'testdouble';
 import {mount, shallow} from 'enzyme';
 import NotchedOutline from '../../../packages/notched-outline/index';
+import {MDCNotchedOutlineAdapter} from '@material/notched-outline/adapter';
 
 suite('NotchedOutline');
 
+function getAdapter(foundation: any): MDCNotchedOutlineAdapter {
+  // @ts-ignore
```
Comment (moog16) [inline test/unit/notched-outline/index.test.tsx]: what are you ignoring here. please comment.

File: test/unit/notched-outline/index.test.tsx [hunk 8]
```diff
@@ -100,57 +87,21 @@ test(
...
-    wrapper.instance().foundation_.adapter_.getHeight(),
-    outlineHeight
-  );
-  div.remove();
-});
-
 test('#adapter.addClass adds class to classList', () => {
   const wrapper = shallow<NotchedOutline>(<NotchedOutline />);
-  wrapper.instance().foundation_.adapter_.addClass('test-class-name');
+  getAdapter(wrapper.instance().foundation_).addClass('test-class-name');
```
Comment (moog16) [inline test/unit/notched-outline/index.test.tsx]: Is this method necessary?

File: test/unit/notched-outline/index.test.tsx [hunk 9]
```diff
@@ -3,29 +3,31 @@ import {assert} from 'chai';
...
 
 test('creates outlineElement_', () => {
   const wrapper = mount<NotchedOutline>(<NotchedOutline />);
   assert.exists(wrapper.instance().outlineElement_.current);
 });
 
-test('creates pathElement_', () => {
-  const wrapper = mount<NotchedOutline>(<NotchedOutline />);
-  assert.exists(wrapper.instance().pathElement_.current);
+test('renders chilren outlineElement_', () => {
```
Comment (moog16) [inline test/unit/notched-outline/index.test.tsx]: children

File: test/unit/notched-outline/index.test.tsx (hunk 7 above)
Comment (gugu) [reply test/unit/notched-outline/index.test.tsx]: adapter_ method is protected in mdc-web, that's how I override it

File: test/unit/notched-outline/index.test.tsx (hunk 8 above)
Comment (gugu) [reply test/unit/notched-outline/index.test.tsx]: it allows to put code with ts-ignore in one place. The adapter_ property is protected in mdc-web

File: test/unit/notched-outline/index.test.tsx (hunk 7 above)
Comment (moog16) [reply test/unit/notched-outline/index.test.tsx]: gotcha - please comment that inline with the ignore. Thanks!

File: packages/notched-outline/index.tsx (hunk 2 above)
Comment (gugu) [reply packages/notched-outline/index.tsx]: I checked this piece of code from https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield

```html
<div class="mdc-text-field mdc-text-field--outlined mdc-text-field--no-label">
  <input type="text" class="mdc-text-field__input" aria-label="Label">
  <div class="mdc-notched-outline">
    <div class="mdc-notched-outline__leading"></div>
    <div class="mdc-notched-outline__trailing"></div>
  </div>
</div>
```

File: packages/notched-outline/index.tsx (hunk 6 above)
Comment (gugu) [reply packages/notched-outline/index.tsx]: without it typescript throws an error

`Property 'foundation_' does not exist on type 'NotchedOutline'.`

Comment (moog16) [reply packages/notched-outline/index.tsx]: I am referring to `outlineElement_: React.RefObject<HTMLDivElement> = React.createRef();`

File: packages/notched-outline/index.tsx (hunk 2 above)
Comment (moog16) [reply packages/notched-outline/index.tsx]: Ok I see that now: https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield#outlined-1

**Comment by moog16** [comment]
@gugu please take a look at the text-field. Maybe we should do floating-label before notched-outline?

**Comment by gugu** [comment]
> @gugu please take a look at the text-field. Maybe we should do floating-label before notched-outline?

I can start with a floating label and we can merge them together

**Comment by gugu** [comment]
Still working on it

**Comment by moog16** [comment]
No worries! There's a lot to do. Take your time :)

**Comment by moog16** [comment]
@gugu - i just merged the floating label updates to feat/mdcweb-typescript-update

**Comment by bartiaco** [comment]
Are there any status updates on this one? We currently are using MDC web and MDC web-react in our project and the version mismatches are causing us conflicts.

Trying to figure out if we should rollback our MDC web components to 0.4X, or try to hack our way around this issue.

**Comment by moog16** [comment]
@gugu was last working on it, and waiting on me for the floating-label changes. We have been tackling other components, but I'm sure we'll get to it soon. Our release will be mid-April so you wouldn't be able to get this change until then.

Are you having conflicts with text-field? If so, maybe just rollback mdcweb text field to 0.4x for the time being.

**Comment by moog16** [comment]
@gugu looks like you started working on this again. LMK when you want me to take a look

**Comment by gugu** [comment]
@moog16 if you can - please take a look. If code looks okay (even if it does not work as expected) - I'll branch it to implement textfield

File: packages/notched-outline/index.tsx [hunk 10]
```diff
@@ -91,49 +89,45 @@ export default class NotchedOutline extends React.Component<
...
         classList.delete(className);
         this.setState({classList});
       },
-      setOutlinePathAttr: (value: string) => {
-        if (this.pathElement_.current) {
-          this.pathElement_.current.setAttribute('d', value);
-        }
+      setNotchWidthProperty: (width: number) => {
+        this.setState({
+          actualWidth: width,
```
Comment (moog16) [inline packages/notched-outline/index.tsx]: I'm confused why there is `actualWidth` and `notchWidth`. `notchWidth` is supposed to come from text-field, and then `actualWidth` is set from the foundation, which adds any padding. Is this correct?

**Comment by moog16** [comment]
@gugu how is text field going?

**Comment by moog16** [comment]
@gugu I'm going to pick up these changes. Hope that is ok.

**Pull request closed without merging**